### PR TITLE
Ref issues/1858: Add test case to cover scenario decode result of DynamicArray<DynamicStruct>

### PR DESCRIPTION
**Pull request by Gelassen**
Ref issues/1858: Add test case to cover scenario decode result of DynamicArray<DynamicStruct>. It confirms issue, but doesn't have solution yet.

### What does this PR do?
Add test case to FunctionReturnDecoder to spot known issue. Ref: https://github.com/web3j/web3j/issues/1858

### Where should the reviewer start?
Touched just several classes. This one probably would be a right start point: 
https://github.com/web3j/web3j/blob/199914cfffda72f3233abd72d4141fdc62b873d5/abi/src/test/java/org/web3j/abi/FunctionReturnDecoderTest.java

### Why is it needed?
Confirm issue in existing codebase, but complete solution would take much more work roughly estimated in 1 man-month work by discussion with former lib developer . Ref: https://github.com/web3j/web3j/issues/1858

Depends on this PR: https://github.com/web3j/web3j/pull/1861

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,5 +1,5 @@
 
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.3-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.3-bin.zip
```
Comment (SweeXordious) [inline gradle/wrapper/gradle-wrapper.properties:2]: Is there a reason to update the gradle version? IMO, this should be handled in a separate PR if needed

File: local.properties [hunk 2]
```diff
@@ -0,0 +1,8 @@
+## This file must *NOT* be checked into Version Control Systems,
```
Comment (SweeXordious) [inline local.properties:1]: As specified here, this file shouldn't be added

File: abi/src/test/java/org/web3j/abi/FunctionReturnDecoderTest.java [hunk 3]
```diff
@@ -1298,4 +1298,40 @@ public void testDecodeTupleOfStaticArrays() {
...
+    /**
+     * Add extra test case to cover discovered edge-case issue
+     *
+     * Ref: https://github.com/web3j/web3j/issues/1858
+     * */
+    @Test
+    public void testDecodeDynamicArrayOfMatches() {
+        String rawInput =
+                "0x"
+                        + "0000000000000000000000000000000000000000000000000000000000000020"  // not sure about this parameter - dynamic array identifier?
```
Comment (SweeXordious) [inline abi/src/test/java/org/web3j/abi/FunctionReturnDecoderTest.java:1311]: yes, this specifies the value `20` in hex which is `32` in decimal which is the offset of the start of the dynamic type

File: abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java [hunk 4]
```diff
@@ -883,4 +890,101 @@ public ArrayStruct(Uint256 id, DynamicArray<Address> addresses) {
...
+                    "\"valueOfFirstUser\": \"${valueOfFirstUser}\", " +
+                    "\"userSecond\": \"${userSecond}\", " +
+                    "\"valueOfSecondUser\" : \"${valueOfSecondUser}\", " +
+                    "\"approvedByFirstUser\" : \"${approvedByFirstUser}\", " +
+                    "\"approvedBySecondUser\" : \"${approvedBySecondUser}\" " +
+                    "}";
+        }
+    }
+
+    public static final Function setArrayOfMatchStructFunction =
```
Comment (SweeXordious) [inline abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java:960]: this function is not tested, better remove it IMO

**Review by SweeXordious — COMMENTED** [review]
Thanks for giving this a shot. Is the test passing locally? or as expected, it's failing?

**Comment by Gelassen** [comment]
> Thanks for giving this a shot. Is the test passing locally? or as expected, it's failing?

It fails as expected. Although the test case shows a different offset in the error, different from what I saw in my personal project.

File: gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (Gelassen) [reply gradle/wrapper/gradle-wrapper.properties:2]: It is possible to find a workaround, but it would be better to update it. There is a compatibility issue between ```gradle``` and ```java``` versions, without resolving it I am not able even a compile a project. I don't use the latest ```JDK```, just one which is comfortable to work with on the rest of my projects. I believe it is a time to migrate to more recent ```gradle``` version. 

I will submit a new PR with change in gradle version, but change in this PR code will be left as is.

File: abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java [hunk 5]
```diff
@@ -117,6 +118,12 @@ public class AbiV2TestFixture {
     public static final String FUNC_setArrayOfStructWithArraysFunction =
             "setArrayOfStructWithArraysFunction";
 
+    // TODO Shall we reconsider const naming convention and unified constants here?
```
Comment (Gelassen) [inline abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java:121]: Please note ```//TODO``` I left here. What would be your opinion on this?

File: gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (Gelassen) [reply gradle/wrapper/gradle-wrapper.properties:2]: Please, here is a separate PR with this change: https://github.com/web3j/web3j/pull/1861

File: abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java (hunk 5 above)
Comment (SweeXordious) [reply abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java:121]: Certainly you could, and that way is more readable, if you open a PR, I would :+1: for it, but would need one of the maintainers to approve it and approve the renaming.

Thanks a lot for the time you're giving the library.

Comment (Gelassen) [reply abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java]: You are welcome. Thank you for all your work, otherwise I have to start all work on the similar library since the beginning. :)

Comment (SweeXordious) [reply abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java]: Generally, if you already have your smart contract written and your structs defined, you could either fork the code and tweak it to work for your case, or you could even write an encoder/decoder for the specific structs you have from scratch easily.

Comment (Gelassen) [reply abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java]: Forking a codebase would lead to constant sync with origin repo. In case of multiple commits it means possible conflict resolution and parallel library maintaining. I thought about this, but left it as a last resort.

Thank you for point out on flexibility in customisation of type encoder\decoder. I will look at this. If you have this documented somewhere, it would be good to see links too.

Comment (SweeXordious) [reply abi/src/test/java/org/web3j/abi/AbiV2TestFixture.java]: There is unfortunately no documentation for this, I haven't checked in a while tho.

But generally, I based my work on [this](https://docs.soliditylang.org/en/v0.8.17/abi-spec.html). If you manage to understand how the encoding/decoding works for static types and dynamic types, you could easily write an encoder/decoder specific to certain structs. That is because you already know what to expect and the fields offsets.

To test your knowledge on the offsets, check the tests scenarios [here](https://github.com/web3j/web3j/blob/master/abi/src/test/java/org/web3j/abi/FunctionReturnDecoderTest.java). They provide encodings + the corresponding structs. That should be a good starting point.
